### PR TITLE
enablement due to dbt-utils' "dispatchifiction"

### PR DESCRIPTION
**Pull request by swanderz**
With the two PRs below, we can now do the following two things:

### support as many macros as we can.

as of now, all 39 of dbt-utils's macros are fully working except for the 15 listed below:
- `sql/groupby.sql` (will never be supported)
- `insert_by_period` materialization
- `date_spine`
- `test_mutually_exclusive_ranges()` #18 
- all the `url` macros (`get_url_host`, `get_url_path`, `get_url_host()` #21
- `generate_series()`
- `get_relations_by_pattern()`
- `get_relations_by_prefix_and_union()`
- `union_relations()`

additionally the following macros work on Azure SQL & SQL Server, but due to a [synapse bug](https://github.com/dbt-msft/dbt-synapse/issues/36) are disabled, because the ITs don't work. 
- `dateadd()`
- `datediff()`
- `split_part()`
- `last_day()`
- `test-hash()`

### `dbt-utils`'s integration testing

The goal of getting this repo working with dbt-utils' integration testing was to make it very easy to:
- test that our shimmed macros, and
- keep said macros up-to-date as dbt-utils changes.

One challenge with using dbt-utils' integration testing was that it is a bit of a [RepRap](https://en.wikipedia.org/wiki/RepRap_project) in that it includes tests (schema and data) that are needed to evaluate whether the other macros are working. So in some cases, a macro works in TSQL with only a small change, but we had to disable the model corresponding to the macro in the `dbt_project.yml` anyway because there are tests used on the macro that weren't compatible with TSQL. 

dependent, now-merged PRs:
- https://github.com/fishtown-analytics/dbt-utils/pull/310
- https://github.com/fishtown-analytics/dbt-utils/pull/312

### CODE REVIEW

File: macros/dbt_utils/schema_tests/mutually_exclusive_ranges.sql [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            false
+        ) as lower_bound_less_than_upper_bound,
+
+        -- For each record: upper_bound {{ allow_gaps_operator }} the next lower_bound.
+        -- Coalesce it to handle null cases for the last record.
+        coalesce(
+            upper_bound {{ allow_gaps_operator }} next_lower_bound,
+            is_last_record,
+            false
+        ) as upper_bound_{{ allow_gaps_operator_in_words }}_next_lower_bound
```
Comment (swanderz) [inline macros/dbt_utils/schema_tests/mutually_exclusive_ranges.sql:67]: @NandanHegde15 @jkock @alittlesliceoftom when any of you have a chance do you mind looking at converting these coalesce statements into `NULLIF`s or `CASE WHEN`s? Also `false` needs to be replaced with `0` bc TSQL doesn't have a true & false. Here's[ a gist of the compiled SQL](https://gist.github.com/swanderz/6ebedfb73554d2e6d6759d4bb4f3770e) that's error-ing out. If you log into the test db (`dbt-msft-serverless-db`), you can debug it.

File: macros/dbt_utils/schema_tests/mutually_exclusive_ranges.sql [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        
+        --TODO turn thesse into null ifs or case whens...
+
+        -- For each record: lower_bound should be < upper_bound.
+        -- Coalesce it to return an error on the null case (implicit assumption
+        -- these columns are not_null)
+        coalesce(
+            lower_bound < upper_bound,
+            false
+        ) as lower_bound_less_than_upper_bound,
```
Comment (swanderz) [inline macros/dbt_utils/schema_tests/mutually_exclusive_ranges.sql:59]: said another way, this logic is:

> - if `lower_bound < upper_bound` is not null:
>     - return the result of the comparison: `true` or `false` (or in TSQL's case: `1` or `0`)
> - else if  `lower_bound < upper_bound` is `NULL`:
>     -  return `false` (i.e. `0`)
though i'm not sure in what scenario `lower_bound < upper_bound` could be `NULL`

**Comment by swanderz** [comment]
@jtcohen6 any idea as to how we can tackle these tests? Maybe wrap the majority of logic into standalone macros that we can write our own versions of?
https://github.com/swanderz/dbt-utils/pull/1/files

**Comment by jtcohen6** [comment]
> @jtcohen6 any idea as to how we can tackle these tests? Maybe wrap the majority of logic into standalone macros that we can write our own versions of?
> https://github.com/swanderz/dbt-utils/pull/1/files

@swanderz We could abstract those in `{{ select_one() }}` `{{ limit_zero() }}` macros, but your proposed changes also seem like reasonable rewrites to me that wouldn't break any mainline behavior. Want to add them to https://github.com/fishtown-analytics/dbt-utils/pull/310?

**Comment by swanderz** [comment]
> Want to add them to [fishtown-analytics/dbt-utils#310](https://github.com/fishtown-analytics/dbt-utils/pull/310)?

sure, though i'm leery of the size of that PR, so maybe in a follow-on 🤷 ?

**Comment by jtcohen6** [comment]
> > Want to add them to [fishtown-analytics/dbt-utils#310](https://github.com/fishtown-analytics/dbt-utils/pull/310)?
> 
> sure, though i'm leery of the size of that PR, so maybe in a follow-on 🤷 ?

Feels closely related, and that PR isn't too big (talking about "t-sql hacks that shouldn't break mainline behavior," not "dispatch-ify ALL THE MACROS!")

File: integration_tests/dbt_utils/dbt_project.yml [hunk 3]
```diff
@@ -17,46 +17,26 @@ clean-targets:         # directories to be removed by `dbt clean`
     - "dbt_modules"
     
 vars:
-  dbt_utils_dispatch_list: ['tsql_utils']
+  dbt_utils_dispatch_list: ['tsql_utils', 'dbt_utils']
```
Comment (swanderz) [inline integration_tests/dbt_utils/dbt_project.yml:20]: added this as a hail mary.... so close now!
```
almost there... do I have to add a var to `tsql-utils/integration_tests/dbt_utils/packages.yml`?
```
  In dispatch: No macro named 'limit_zero' found
      Searched for: 'dbt_utils.sqlserver__limit_zero', 'dbt_utils.default__limit_zero'
```
```

Comment (jtcohen6) [reply integration_tests/dbt_utils/dbt_project.yml:20]: Good call, this should actually be the `name` of the project (`integration_tests/dbt_utils/dbt_project.yml`) that defines `sqlserver__limit_zero` (`integration_tests/dbt_utils/macros/limit_zero.yml`):
```suggestion
  dbt_utils_dispatch_list: ['tsql_utils', 'tsql_utils_dbt_utils_integration_tests']
```

File: integration_tests/dbt_utils/dbt_project.yml [hunk 4]
```diff
@@ -30,12 +30,19 @@ models:
...
       test_get_relations_by_pattern: *disabled
       test_get_relations_by_prefix_and_union: *disabled
       test_groupby: *disabled # TSQL doesn't let you group by column numbers
+      # BROKEN DUE TO SYNAPSE SEED LOADING EMPTY CSV VALS AS EMPTY STRINGS
+      # see: https://github.com/dbt-msft/dbt-synapse/issues/36
+      test_dateadd: *not-synapse
+      test_datediff: *not-synapse
+      test_hash: *not-synapse
+      test_last_day: *not-synapse
+      test_union: *not-synapse
```
Comment (swanderz) [inline integration_tests/dbt_utils/dbt_project.yml:45]: @jtcohen6 I thought that this config change would  take care of the below. but it isn't.
```
Completed with 4 errors and 0 warnings:

Failure in test assert_equal_test_dateadd_actual__expected (models/cross_db_utils/schema.yml)
  Got 1 result, expected 0.

  compiled SQL at target/compiled/dbt_utils_integration_tests/models/cross_db_utils/schema.yml/schema_test/assert_equal_test_dateadd_actual__expected.sql

Failure in test assert_equal_test_datediff_actual__expected (models/cross_db_utils/schema.yml)
  Got 2 results, expected 0.

  compiled SQL at target/compiled/dbt_utils_integration_tests/models/cross_db_utils/schema.yml/schema_test/assert_equal_test_datediff_actual__expected.sql

Failure in test assert_equal_test_hash_actual__expected (models/cross_db_utils/schema.yml)
  Got 1 result, expected 0.

  compiled SQL at target/compiled/dbt_utils_integration_tests/models/cross_db_utils/schema.yml/schema_test/assert_equal_test_hash_actual__expected.sql

Failure in test assert_equal_test_last_day_actual__expected (models/cross_db_utils/schema.yml)
  Got 1 result, expected 0.

  compiled SQL at target/compiled/dbt_utils_integration_tests/models/cross_db_utils/schema.yml/schema_test/assert_equal_test_last_day_actual__expected.sql
```

Comment (jtcohen6) [reply integration_tests/dbt_utils/dbt_project.yml:45]:
```
[WARNING]: Configuration paths exist in your dbt_project.yml file which do not apply to any resources.
There are 4 unused configuration paths:
- models.dbt_utils_integration_tests.sql.test_dateadd
- models.dbt_utils_integration_tests.sql.test_last_day
- models.dbt_utils_integration_tests.sql.test_hash
- models.dbt_utils_integration_tests.sql.test_datediff
```
The reason is that these are located under `cross_db_utils`, not `sql`, in [the file structure](https://github.com/fishtown-analytics/dbt-utils/tree/master/integration_tests/models/cross_db_utils). So you want to move these higher up:
```yml
models:
  dbt_utils_integration_tests:
    cross_db_utils:
      test_split_part: &not-synapse
        +enabled: "{{ target.name != 'synapse' }}"
      test_dateadd: *not-synapse
      test_datediff: *not-synapse
      test_hash: *not-synapse
      test_last_day: *not-synapse
```

Comment (swanderz) [reply integration_tests/dbt_utils/dbt_project.yml]: 🤦

**Comment by alittlesliceoftom** [comment]
Could I be really boring and request: 
1. Slightly more description of the PR - it seems we're (a) enabling a bunch more things due to the other PRs and (b) also adding some new functionality e.g. time datatypes. Presumably these are fixes in response to tests we can now run due to (a)?
2. resolve the merge conflict

**Comment by swanderz** [comment]
> Could I be really boring and request:
> 
>     1. Slightly more description of the PR - it seems we're (a) enabling a bunch more things due to the other PRs and (b) also adding some new functionality e.g. time datatypes. Presumably these are fixes in response to tests we can now run due to (a)?
> 
>     2. resolve the merge conflict

done. and done.

**Comment by swanderz** [comment]
@mikaelene  what do you think?

File: macros/dbt_utils/cross_db_utils/hash.sql [hunk 5]
```diff
@@ -1,5 +1,5 @@
 {% macro sqlserver__hash(field) %}
-    hashbytes('md5', {{field}})
+    convert(varchar(50), hashbytes('md5', {{field}}), 2)
```
Comment (alittlesliceoftom) [inline macros/dbt_utils/cross_db_utils/hash.sql:2]: @swanderz just noting that this will limit hashable field size to 50 (bytes not chars I think). Is this intended behaviour? Could use string?

File: .gitmodules [hunk 6]
```diff
@@ -1,3 +1,4 @@
 [submodule "dbt-utils"]
 	path = dbt-utils
 	url = https://github.com/fishtown-analytics/dbt-utils
+	branch = master
```
Comment (alittlesliceoftom) [inline .gitmodules:4]: Just noting this may result in changes in dbt-utils breaking things on our side without an explicit commit as not pegged to a certain version.

**Review by alittlesliceoftom — APPROVED** [review]
Hey @swanderz , 

This all looks really good. I haven't marked last 4 (test files) as 'viewed' as I didn't fully understand them as I haven't worked with those macros. But I definitely think this is all good to pull in.

**Comment by alittlesliceoftom** [comment]
> Hey @swanderz ,
> 
> This all looks really good. I haven't marked last 4 (test files) as 'viewed' as I didn't fully understand them as I haven't worked with those macros. But I definitely think this is all good to pull in.

You might consider adding some docstrings to the tests to help others (and yourself when you forget what you did) to understand them as I think they're quite complex to reason about if looking at from fresh

File: macros/dbt_utils/cross_db_utils/hash.sql (hunk 5 above)
Comment (mikaelene) [reply macros/dbt_utils/cross_db_utils/hash.sql:2]: I looked at this as well. I had the same in the old sqlserver-utils package. Maybe we can have varchar(max)? Most cases will be hashing one column, but if you want to look for diffs, like watching for changes, you may include a lot of columns making the hashes longer then 50

**Comment by mikaelene** [comment]
> @mikaelene what do you think?

I think this looks great on an overall level. Don't have time to go through all but really great work!

**Comment by swanderz** [comment]
> > Hey @swanderz ,
> > This all looks really good. I haven't marked last 4 (test files) as 'viewed' as I didn't fully understand them as I haven't worked with those macros. But I definitely think this is all good to pull in.
> 
> You might consider adding some docstrings to the tests to help others (and yourself when you forget what you did) to understand them as I think they're quite complex to reason about if looking at from fresh

Some context about the bottom 3`schema_test/` macros you refer to. one of the coolest things about `dbt-utils` are their integration tests, which are really unit tests. For many macros, there's two seed models: an input and an output. The testing process is that the macro is run on the input model and then, the output is compared to output seed.

The challenge is that some arguments defined in schema tests for some testing models in `dbt-utils/integration_tests/models/schema_tests/schema.yml` use things that are not supported by TSQL. Things like `<>` for not equal, or `something = false` (TSQL doesn't have `true` or `false` values. So rather than making a PR to dbt-utils and changing the args, it's better to just capture them inside of our dispatched macros and swap them for the TQL equivalent.

TL;DR the changes are just to make the integration testing work, and shouldn't ever affect end users unless I'm missing something
